core/src/main/java/org/elasticsearch/percolator/PercolatorQuery.java

```diff
@@ -0,0 +1,173 @@
+                return innerWeight.getValueForNormalization();
+            }
+
+            @Override
+            public void normalize(float v, float v1) {
+                innerWeight.normalize(v, v1);
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext leafReaderContext) throws IOException {
+                final Scorer approximation = innerWeight.scorer(leafReaderContext);
+                if (approximation == null) {
+                    return null;
+                }
+
+                final LeafReader leafReader = leafReaderContext.reader();
+                return new Scorer(this) {
+
+                    @Override
+                    public TwoPhaseIterator asTwoPhaseIterator() {
+                        return new TwoPhaseIterator(approximation) {
+                            @Override
+                            public boolean matches() throws IOException {
+                                return matchDocId(approximation.docID(), leafReader);
+                            }
+                        };
+                    }
+
+                    @Override
+                    public float score() throws IOException {
+                        return approximation.score();
+                    }
+
+                    @Override
+                    public int freq() throws IOException {
+                        return approximation.freq();
+                    }
+
+                    @Override
+                    public int docID() {
+                        return approximation.docID();
+                    }
+
+                    @Override
+                    public int nextDoc() throws IOException {
+                        return advance(approximation.docID() + 1);
+                    }
+
+                    @Override
+                    public int advance(int target) throws IOException {
+                        for (int docId = approximation.advance(target); docId < NO_MORE_DOCS; docId = approximation.advance(docId + 1)) {
+                            if (matchDocId(docId, leafReader)) {
+                                return docId;
+                            }
+                        }
+                        return NO_MORE_DOCS;
+                    }
+
+                    @Override
+                    public long cost() {
+                        return approximation.cost();
+                    }
+                };
+            }
+        };
+    }
+
+    boolean matchDocId(int docId, LeafReader leafReader) throws IOException {
+        SingleFieldsVisitor singleFieldsVisitor = new SingleFieldsVisitor(UidFieldMapper.NAME);
+        leafReader.document(docId, singleFieldsVisitor);
+        BytesRef percolatorQueryId = new BytesRef(singleFieldsVisitor.uid().id());
+        Query percolatorQuery = percolatorQueries.get(percolatorQueryId);
+        if (percolatorQuery != null) {
+            Lucene.exists(percolatorIndexSearcher, percolatorQuery, collector);
```

Comment:
I'm ok with creating an instance per scorer. The reason I though is was safe is that we don't execute a query in parallel of the segments and the `Lucene.exists(...)` method invokes the reset method of the collector.